src/core/obj.js

```diff
@@ -595,6 +586,129 @@ var Catalog = (function CatalogClosure() {
     }
   };
 
+  /**
+   * Helper function used to parse the contents of destination dictionaries.
+   * @param {Dict} destDict - The dictionary containing the destination.
+   * @param {Object} resultObj - The object where the parsed destination
+   *   properties will be placed.
+   * @param {string} docBaseUrl - (optional) The document base URL that is used
+   *   when attempting to recover valid absolute URLs from relative ones.
+   */
+  Catalog.parseDestDictionary = function Catalog_parseDestDictionary(params) {
+    // Lets URLs beginning with 'www.' default to using the 'http://' protocol.
+    function addDefaultProtocolToUrl(url) {
+      if (isString(url) && url.indexOf('www.') === 0) {
+        return ('http://' + url);
+      }
+      return url;
+    }
+    // According to ISO 32000-1:2008, section 12.6.4.7, URIs should be encoded
+    // in 7-bit ASCII. Some bad PDFs use UTF-8 encoding, see Bugzilla 1122280.
+    function tryConvertUrlEncoding(url) {
+      try {
+        return stringToUTF8String(url);
+      } catch (e) {
+        return url;
+      }
+    }
+
+    var destDict = params.destDict;
+    if (!isDict(destDict)) {
+      warn('Catalog_parseDestDictionary: "destDict" must be a dictionary.');
+      return;
+    }
+    var resultObj = params.resultObj;
+    if (typeof resultObj !== 'object') {
+      warn('Catalog_parseDestDictionary: "resultObj" must be an object.');
+      return;
+    }
+    var docBaseUrl = params.docBaseUrl || null;
+
+    var action = destDict.get('A'), url, dest;
+    if (isDict(action)) {
+      var linkType = action.get('S').name;
+      switch (linkType) {
+        case 'URI':
+          url = action.get('URI');
+          if (isName(url)) {
+            // Some bad PDFs do not put parentheses around relative URLs.
+            url = '/' + url.name;
+          } else if (url) {
+            url = addDefaultProtocolToUrl(url);
+          }
+          // TODO: pdf spec mentions urls can be relative to a Base
+          // entry in the dictionary.
+          break;
+
+        case 'GoTo':
+          dest = action.get('D');
+          break;
+
+        case 'GoToR':
+          var urlDict = action.get('F');
+          if (isDict(urlDict)) {
+            // We assume that we found a FileSpec dictionary
+            // and fetch the URL without checking any further.
+            url = urlDict.get('F') || null;
+          } else if (urlDict) {
```

Comment:
We lost `isString(urlDict)` here. Below we expect it to be a string (it can be array or other type).